api_app/connectors_manager/connectors/opencti.py

```diff
@@ -0,0 +1,171 @@
+from tests.mock_utils import patch, if_mock_connections
+from api_app import helpers
+from api_app.connectors_manager import classes
+
+
+INTELOWL_OPENCTI_TYPE_MAP = {
+    "ip": {
+        "v4": "ipv4-addr",
+        "v6": "ipv6-addr",
+    },
+    "domain": "domain-name",
+    "url": "url",
+    # type hash is combined with file
+    "generic": "x-opencti-text",  # misc field, so keeping text
+    "file": "file",  # hashes: md5, sha-1, sha-256
+}
+
+
+class OpenCTI(classes.Connector):
+    def set_params(self, params):
+        self.ssl_verify = params.get("ssl_verify", True)
+        self.tlp = params.get(
+            "tlp", {"type": "white", "color": "#FFFFFF", "x_opencti_order": 0}
+        )
+        self.proxies = params.get("proxies", {})
+        self.__url_name = self._secrets["url_key_name"]
+        self.__api_key = self._secrets["api_key_name"]
+
+    def get_observable_type(self) -> str:
+        if self._job.is_sample:
+            obs_type = INTELOWL_OPENCTI_TYPE_MAP["file"]
+        elif self._job.observable_classification == "hash":
+            matched_hash_type = helpers.get_hash_type(self._job.observable_name)
+            if matched_hash_type in [
+                "md5",
+                "sha-1",
+                "sha-256",
+            ]:  # sha-512 not supported
+                obs_type = INTELOWL_OPENCTI_TYPE_MAP["file"]
+            else:
+                obs_type = INTELOWL_OPENCTI_TYPE_MAP["generic"]  # text
+        elif self._job.observable_classification == "ip":
+            ip_version = helpers.get_ip_version(self._job.observable_name)
+            if ip_version == 4 or ip_version == 6:
+                obs_type = INTELOWL_OPENCTI_TYPE_MAP["ip"][f"v{ip_version}"]  # v4/v6
+            else:
+                obs_type = INTELOWL_OPENCTI_TYPE_MAP["generic"]  # text
+        else:
+            obs_type = INTELOWL_OPENCTI_TYPE_MAP[self._job.observable_classification]
+
+        return obs_type
+
+    def generate_observable_data(self) -> dict:
+        observable_data = {"type": self.get_observable_type()}
+        if self._job.is_sample:
+            observable_data["name"] = self._job.file_name
+            observable_data["hashes"] = helpers.generate_hashes(self.job_id)
+        elif (
+            self._job.observable_classification == "hash"
+            and observable_data["type"] == "file"
+        ):
+            # add hash instead of value
+            matched_type = helpers.get_hash_type(self._job.observable_name)
+            observable_data["hashes"] = {matched_type: self._job.observable_name}
+        else:
+            observable_data["value"] = self._job.observable_name
+
+        return observable_data
+
+    def run(self):
+        # set up client
+        opencti_instance = pycti.OpenCTIApiClient(
+            url=self.__url_name,
+            token=self.__api_key,
+            ssl_verify=self.ssl_verify,
+            proxies=self.proxies,
+        )
+
+        # Create author (if it doesn't exist)
```

Comment:
I need the uuid of an author for further steps.

Reply:
if we save the uuid in a new table of the DB at the time of the first creation? and we try to retrieve it at the time of the analysis? If we do not find it, we can just follow the normal flow.

Reply:
Something like `cached_property` could work. Similar to how we use a session in PyIntelOwl ([reference](https://github.com/intelowlproject/pyintelowl/blob/681096d37690a63c26e133956e7036f04295ec43/pyintelowl/pyintelowl.py#L33-L50)).

Reply:
`cached_property` won't be useful as it would stay only with the instance and these are already called once per instance.
We need something more persistent.